examples/mongocxx/document_validation.cpp

```diff
@@ -0,0 +1,75 @@
+#include <iostream>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/stdx/string_view.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/exception/write.hpp>
+#include <mongocxx/instance.hpp>
+#include <mongocxx/stdx.hpp>
+#include <mongocxx/uri.hpp>
+
+using bsoncxx::builder::stream::document;
+using bsoncxx::builder::stream::open_document;
+using bsoncxx::builder::stream::close_document;
+using mongocxx::stdx::string_view;
+using mongocxx::collection;
+using mongocxx::validation_criteria;
+
+int main(int, char**) {
+    mongocxx::instance inst{};
+    mongocxx::client conn{mongocxx::uri{}};
+
+    auto db = conn["test"];
+
+    // Create a collection with document validation enabled.
+    {
+        // @begin: cpp-create-collection-with-document-validation
+        validation_criteria validation;
+        validation.level(validation_criteria::validation_level::k_strict);
+        validation.action(validation_criteria::validation_action::k_error);
+
+        // Add a validation rule: all zombies need to eat some brains.
+        document rule;
+        rule << "brains" << open_document << "$gt" << 0 << close_document;
+        validation.rule(rule.extract());
+
+        mongocxx::options::create_collection opts;
+        opts.validation_criteria(validation);
+
+        // Clean up any old collections with this name
+        if (db.has_collection("zombies")) {
+            db["zombies"].drop();
+        }
+
+        collection zombies = db.create_collection("zombies", opts);
+
+        try {
+            // Insert a document passing validation
+            document betty;
+            betty << "name"
+                  << "Bloody Betty"
+                  << "brains" << 3;
+            auto res = zombies.insert_one(betty.extract());
+
+            std::cout << "Bloody Betty passed document validation!" << std::endl;
+
+            // Insert a document failing validation
+            document fred;
+            fred << "name"
+                 << "Undead Fred"
+                 << "brains" << 0;
+
+            // Inserting a failing document should throw
+            auto res2 = zombies.insert_one(fred.extract());
+
+            std::cout << "UNREACHABLE." << std::endl;
```

Comment:
I think I should change the message to "Your server does not support document validation," because we would end up here if that were the case, and it wouldn't be the driver's fault